# PepTransDB
PepTransDB (Peptides in Transcripts Database, (Peptide Encoding Transcripts Database, https://www.maherlab.com/peptransdb) is a comprehensive integrative analyses of mass spectrometry-based proteomics and transcriptomic sequencing data from greater than 1,000 patients across ten cancer types. The resource has 21,027 novel peptides derived from 9,318 long noncoding RNAs (lncRNAs). We also exploited open reading frames overlapping the backspliced region of circular RNAs (circRNAs) to identify 3,515 peptides that are uniquely derived from 3,002 circRNAs and not their corresponding linear RNAs. We hope this pan-cancer proteogenomic analysis will serve as a resource for evaluating the coding potential of lncRNAs and circRNAs that could aide future mechanistic studies exploring their function in cancer. Here we provide scripts and steps used to generate the data.  


## proteogenomic search and downstream analysis
1. The main steps used to generate the initial set of peptides is in  pipeline.sh, a bash script that utilizes msFragger, peptideProphet and proteinProphet to generate .psm and .protein files for downstream analysis.   
 
Usage:   
```sh pipeline.sh <path_to_mzML_files> <database_file> <parameter_file>```  
Args:   
```<path_to_mzML_files>``` = path to the directory with mzML files for each cancer.  
```<database_file>``` = database of target sequences with reverse decoy sequences with .fasta extension.  
```<parameter_file>``` = parameter file formatted in the format specified by msFragger (2019) and with .params extension.  

2. To parse the raw output from ```pipeline.sh```, ```parse_proteogenomic_output.r``` is used to filter peptides mapped to decoys and retrieve mapped proteins for each cancer.   
Usage:  
```Rscript parse_proteogenomic_output.r <path_to_psm_files>```  
Args:  
```<path_to_psm_files>``` = location of PSM files generated by pipeline.sh  


## Steps to obtain ORFs for the database  
1. Find sORFs using EMBOSS's getORF  
```getorf -sequence lnc_input_mRNA.fa -minsize 100 -outseq lnc_ORF.fa -table 1 -find 1 -noreverse -nomethionine```  
2. Find cORFs using EMBOSS's getORF  
```getorf -sequence circ_input_mRNA.fa -minsize 200 -outseq circ_ORF.fa -table 1 -find 1 -noreverse -nomethionine```  
2. Calculate coordinates of short ORFs generated using EMBOSS's getORF, ```sorf_to_genome.py``` is used utilizing the genomic locations of transcripts (hg38).  
```corf_to_genome.py``` calculates coordinates of circular ORFs from EMBOSS's getORF utilizing the genomic locations of transcriptt (hg38).  
Usage:  
```python corf_to_genome.py <ORF file> <exon file>```  
```python sorf_to_genome.py <ORF file> <exon file>```  
Args:  
```<ORF file>``` = reformatted FASTA file of ORFs from getORF e.g. headers should look like this: >ENST00000456328.2 7 - 63 not >ENST00000456328.2_1 [7 - 63]  
```<exon file>``` = contains exon sequence coordinates of transcripts ran through getORF  
columns (0 based numbering) 0:chromosome, 1:gene name, 2:transcript ID, 3:backsplice identifier, 4:strand, 5:exons (format start;stop,start;stop etc. )  

3. Generate PhyloCSF scores  
```bigWigAverageOverBed https://data.broadinstitute.org/compbio1/PhyloCSFtracks/hg38/latest/PhyloCSF+1.bw -minMax ORF.bed phylocsf_frame1```  
```bigWigAverageOverBed https://data.broadinstitute.org/compbio1/PhyloCSFtracks/hg38/latest/PhyloCSF+3.bw -minMax ORF.bed phylocsf_frame3```  
```bigWigAverageOverBed https://data.broadinstitute.org/compbio1/PhyloCSFtracks/hg38/latest/PhyloCSF+2.bw -minMax ORF.bed phylocsf_frame2```  


## Generatation of junction sequences for CircRNA analysis  
1. Extract coordinates of circRNAs from miOncoCirc_v0.1.release.txt ( https://mioncocirc.github.io/download/) using samples info in circRNAData and exons coordinates in bed format. 
```Rscript extractSeq_miOncoCirc.r```  
2. Retrieve sequences of junction exons  
```sh getfasta_junction_miOncoCirc.sh```  
3. Concatinate junction exon sequences  
```sh getfasta_junction_miOncoCirc.sh```  
4. Get normalized expression matrices for circRNAs using miOncoCirc_seq_depth available in circRNAData  
```Rscript norm_exp_matrix_miOncoCirc.r```  
 
## How to cite PepTransDB?  
Coming soon...  

## Contact  
gothoum (at) wustl (dot) edu  
